feat(provider/cohere): reasoning model support

gr2m opened the pull request:
## Background

Cohere introduced reasoning with their latest model `command-a-reasoning-08-2025`:
https://docs.cohere.com/docs/reasoning

## Summary

Implements handling for reasoning in `CohereChatLanguageModel` for both `generateText()` and `streamText()` as well as provider options to configure the reasoning.

## Manual Verification

1. [examples/ai-core/src/generate-text/cohere-reasoning.ts
](https://github.com/vercel/ai/blob/13ac81da66886f8080c9f1bcf6ddfb5bf6f6dcae/examples/ai-core/src/generate-text/cohere-reasoning.ts)
2. [examples/ai-core/src/stream-text/cohere-reasoning.ts
](https://github.com/vercel/ai/blob/13ac81da66886f8080c9f1bcf6ddfb5bf6f6dcae/examples/ai-core/src/stream-text/cohere-reasoning.ts)

## Tasks

- [x] Tests have been added / updated (for bug fixes / features)
- [x] Documentation has been added / updated (for bug fixes / features)
- [x] A _patch_ changeset for relevant packages has been added (for bug fixes / features - run `pnpm changeset` in the project root)
- [x] Formatting issues have been fixed (run `pnpm prettier-fix` in the project root)
- [x] I have reviewed this pull request (self-review)

## Future Work

🤷🏼 

## Related Issues

closes #8239

gr2m committed:
docs: `generatText()` example

gr2m committed:
doc: `streamText()` example

gr2m committed:
docs: changeset

gr2m committed:
test

gr2m committed:
feat

gr2m committed:
docs: remove debugger from example

gr2m committed:
docs(example): log request body and response content

gr2m committed:
docs: "reasoning" section in provider guide

gr2m committed:
fixup! test

gr2m committed:
fixup! docs: `generatText()` example

gr2m committed:
fixup! feat

gr2m committed:
fixup! docs: `generatText()` example

gr2m reviewed (commented):
I think different providers reasoning works in different ways. In the case of Coherer, the streaming sends two `content-start` and `content-end` events if reasoning is present, first for reasoning, then for text.

lgrammel commented on packages/cohere/src/cohere-chat-language-model.test.ts:
```diff
@@ -45,6 +47,14 @@ describe('doGenerate', () => {
     generation_id?: string;
     headers?: Record<string, string>;
   }) {
+    const content = [];
+    if (typeof text === 'string') {
+      content.push({ type: 'text', text });
+    }
+    if (typeof thinking === 'string') {
+      content.push({ type: 'thinking', thinking });
+    }
```
i would have passed content[] instead and removed txt to match their api more closesly

gr2m replied:
you mean this?

```diff
-    text?: string | null;
-    thinking?: string | null;
+   content: Array<{type:'text', text: string} | {type:'thinking', thinking: string}>;
```

gr2m replied:
ah yes just saw your other comments

lgrammel commented on packages/cohere/src/cohere-chat-language-model.ts:
```diff
@@ -108,6 +121,14 @@ export class CohereChatLanguageModel implements LanguageModelV2 {
         // documents for RAG:
         ...(cohereDocuments.length > 0 && { documents: cohereDocuments }),
+
+        // reasoning
+        ...(cohereOptions.thinking && {
+          thinking: {
+            type: cohereOptions.thinking.type ?? 'enabled',
+            token_budget: cohereOptions.thinking.tokenBudget,
+          },
+        }),
```
in ai sdk 6 it might be worth revisiting default, i.e. should we go with provider defaults or with what we think are recommended settings that cut down time to get started for users

gr2m replied:
Shall we keep track of thoughts like this somewhere? Shall we add a `TODO AI SDK 6` comment?

gr2m committed:
Merge branch 'main' into gr2m/8239-cohere-command-a-reasoning-model-support

gr2m commented:
I tried to finish this up but now run into an error when running the example. I'm looking into it

gr2m committed:
Update packages/cohere/src/cohere-chat-options.ts

Co-authored-by: vercel[bot] <35613825+vercel[bot]@users.noreply.github.com>

gr2m commented on packages/cohere/src/cohere-chat-options.ts:
```diff
@@ -12,3 +15,22 @@ export type CohereChatModelId =
+  thinking: z
+    .object({
+      type: z.enum(['enabled', 'disabled']).optional(),
+      tokenBudget: z.number().optional(),
```
that's okay, I simplified it for coherence with similar schema definitions from other models.

gr2m committed:
test: make mocking futureproof

addresses https://github.com/vercel/ai/pull/8279#discussion_r2299914271 and https://github.com/vercel/ai/pull/8279#discussion_r2299916765

gr2m committed:
refactor

addresses https://github.com/vercel/ai/pull/8279#discussion_r2299920544

gr2m commented on packages/cohere/src/cohere-chat-language-model.ts:
```diff
@@ -272,10 +319,20 @@ export class CohereChatLanguageModel implements LanguageModelV2 {
               }
 
               case 'content-end': {
+                if (isActiveReasoning) {
+                  controller.enqueue({
+                    type: 'reasoning-end',
+                    id: String(value.index),
+                  });
+                  isActiveReasoning = false;
+                  return;
+                }
+
                 controller.enqueue({
                   type: 'text-end',
                   id: String(value.index),
                 });
```
Cohere uses `content-start` and `content-end` for both reasoning and result text. I'm not sure how the implementation could mix up the two?

gr2m committed:
style: prettier

gr2m commented on packages/cohere/src/cohere-chat-language-model.ts:
```diff
@@ -255,6 +284,15 @@ export class CohereChatLanguageModel implements LanguageModelV2 {
 
             switch (type) {
               case 'content-start': {
+                if (value.delta.message.content.type === 'thinking') {
```
use object

```suggestion
                const isThinking = value.delta.message.content.type === 'thinking'
                isReasoning[value.index] = isThinking
                if (isThinking) {
```



compare https://github.com/vercel/ai/blob/cdc6b7a612a2f775f5f6109fb80ef79f8e4f4989/packages/anthropic/src/anthropic-messages-language-model.ts#L894

gr2m commented on packages/cohere/src/cohere-chat-language-model.test.ts:
```diff
@@ -760,33 +771,73 @@ describe('doGenerate', () => {
+
+    if (thinkingContent.length > 0) {
+      thinkingEvents.push(
+        `event: content-start\ndata: {"type":"content-start","index":0,"delta":{"message":{"content":{"type":"thinking","thinking":""}}}}\n\n`,
```
verify in original raw chunks that `index` is set to 0 for both thinking and text content
